### ISSUE TITLE
Flower xray_classification tutorial: deterministic loss=nan on training start (RTX 5090 / torch 2.11)

### ISSUE THREAD

**Issue by atriaybagur** [londonaicentre/FLIP#764]
## Summary

The Flower `xray_classification` tutorial produces **deterministic `loss=nan`** the instant training starts, so the model goes to `ERROR` and the e2e smoke fails on the Flower backend. The nvflare xray tutorial trains the **same data on the same GPU** to `RESULTS_UPLOADED`, so this is specific to the Flower training path, not the platform. Surfaced while re-testing the Docker-hardening PR (#501) — it is **not** related to that PR (the training runs in the unhardened fl-client).

## Environment

- `flwr` 1.32.0
- `torch` 2.11.0+cu128, CUDA 12.8
- GPU: **NVIDIA RTX 5090**, compute capability **(12, 0)** = `sm_120` (Blackwell)
- Backend: Flower; app `flwrlabs.standard-app.1.0.0`; tutorial `fl-tutorials/flower/xray_classification/`

## Reproduction

```bash
make restart-fl FL_BACKEND=flower DOCKER_FL_REGISTRY= DOCKER_FL_TAG=dev
make e2e_smoke FL_BACKEND=flower EXTRA_ARGS="--trusts GSTT"
```

Observed in the training client (`trust1-fl-client-net-*`), consistent across two runs:

```
INFO :      Dataset ready: 298 image/label pairs
INFO :      Split: 178 train, 60 val, 60 test.
INFO :      Using device: cuda
INFO :      Starting training pass on device cuda
INFO :      Train batch 1/23 - loss=nan
INFO :      Train batch 11/23 - loss=nan
INFO :      Training pass complete. loss=nan
INFO :      Validation complete. loss=nan
```

→ flip-api flips the model to `ModelStatus.ERROR`; smoke fails with "Model entered ERROR state".

The NaN appears on **training batch 1** (the very first optimizer step), so it is not a data-tail / missing-image effect — the images download fine (imaging-api serves them, dataset builds with 298 pairs).

## Why this looks like a real regression, not the environment

- The **nvflare** xray tutorial (`fl-tutorials/nvflare/image_classification/xray_classification`) trains the same DICOM cohort on the same RTX 5090 to completion (`RESULTS_UPLOADED`) in the same session — no NaN. So the GPU/driver and the data are fine; the divergence is in the Flower app's training code.
- The Flower xray tutorial was GPU-verified on 2026-06-29 during the Flower 1.27→1.32 upgrade (#664), so this is a **recent** regression (env or code drift since then).

## Suspects to check (for the investigation)

1. **Numerics in the Flower app** — learning rate / loss reduction / input normalization in `fl-tutorials/flower/xray_classification/app/task.py` (and `loss_and_metrics.py`, `transforms.py`). Compare against the nvflare tutorial's `trainer.py`, which trains stably.
2. **torch 2.11 + Blackwell (sm_120)** — a matmul/conv/AMP path that returns NaN on `sm_120` with cu128. Worth an isolated forward/backward on a single batch with anomaly detection (`torch.autograd.set_detect_anomaly(True)`) and, as a bisection, `matmul` precision / disabling AMP.
3. **Weight init / label encoding** mismatch between the Flower app and the data loader.

## Notes

- Not a blocker for PR #501 (hardening) — filed separately so the two don't get conflated.
- Investigation happening in an isolated worktree via the standalone tutorial runner (`make -C fl-tutorials run-tutorial TUTORIAL=xray_classification` for the nvflare harness; the Flower app repro is the e2e path above), so it doesn't disturb the shared dev stack.

**Comment by atriaybagur** [comment]
**Root-caused — it's the Flower app's training-loop robustness, not a torch/Blackwell op.** Reproduced in the actual `flower-fl-base:dev` image on the RTX 5090 (sm_120, torch 2.11.0+cu128).

**Ruled out** (all in-image, on the 5090): torch 2.11/Blackwell (856 real-data fwd/bwd + 300 raw conv/bn/matmul iters, cuDNN autotuning, batch 1–16 → **0 NaN**); the app math (full multi-round loop on the real DICOMs, pretrained & random init, many seeds → **stable**, loss→0.05); monai 1.6.0-vs-1.5.2 (loads DenseNet121 to the **same** sha256, identical transform output on all 300 images); Flower `ArrayRecord` weight transfer (**lossless**). torch is **identical** across the flower and nvflare images — only monai/numpy differ, and neither changes the model or transforms. The Flower training code is numerically **byte-for-byte equivalent** to the nvflare tutorial's.

**Cause:** `train_func` (`app/task.py`) has no non-finite-loss guard, so a single NaN-loss batch back-props into every weight via `optimizer.step()` → every later batch is NaN → `np.nanmean` → `loss=nan` for the whole pass → `ERROR`. Reproduced exactly by poisoning one pixel in batch 1 (all 12 batches NaN, weights poisoned). The one deterministic app-code path to a NaN loss is `get_bce_loss` → `loss.sum()/mask.sum()` = **0/0** for a fully-masked batch (every label `-1`).

**Fix** (both applied in an isolated worktree, demonstrated to reproduce-and-resolve):
1. `app/loss_and_metrics.py`: `loss.sum() / mask.sum().clamp_min(1.0)`
2. `app/task.py` `train_func`: skip non-finite-loss batches (`if not torch.isfinite(loss): continue`) and `torch.nn.utils.clip_grad_norm_(model.parameters(), 1.0)` before `optimizer.step()`.

Evidence — same poisoned batch-1, real DenseNet121 + real DICOMs on the 5090:
```
CURRENT loop: per-batch loss=['nan','nan','nan','nan'], NaN batches=12/12, aggregate=nan, weights_poisoned=True  -> ERROR
FIXED   loop: per-batch loss=['skip','0.747','0.865','0.737'], skipped=1/12, aggregate=0.7298, weights_poisoned=False
FIXED   get_bce_loss(all-masked batch): current -> nan   fixed -> 0.0
```

**Caveat / remaining unknown:** I could not reproduce the *first* NaN with the tutorial's own `xrays_mini_300` data (all clean, trains stably). The live trigger is therefore data-side — most likely a **fully-masked batch** (the confirmed 0/0 path, now guarded) or a **degenerate/NaN pixel** in a DICOM served by GSTT's Orthanc via imaging-api. The fix is robust to both. To pin it exactly, add one diagnostic line at batch 1 of the live run and re-run the e2e on the flower backend:
```python
log(INFO, f"batch1 img[min={images.min():.3g} max={images.max():.3g} nan={torch.isnan(images).any()}] "
          f"valid_labels={(labels != -1).sum().item()}/{labels.numel()}")
```
`nan=True` ⇒ served-pixel bug (fix in imaging-api / add a `NaN`-to-num transform); `valid_labels=0` ⇒ fully-masked (the 0/0 path, already fixed by the denominator clamp).

**Parity note:** the identical gaps exist in the **nvflare** tutorial's `trainer.py`/`loss_and_metrics.py` (same `loss.sum()/mask.sum()`, no NaN-guard/grad-clip) — it only escapes today by not hitting a NaN batch on its data. Worth mirroring the fix there.

**Comment by atriaybagur** [comment]
Fix up in #766 (targets `develop`): clamps the loss denominator + skips non-finite-loss batches + grad-clips, applied to both the flower and nvflare xray tutorials. Verified the real edited functions return `0.0` (not NaN) on a fully-masked batch in `flare-fl-base:dev`. The data-side trigger diagnostic noted above is still worth running on the next flower e2e.
